asset-swapper: Punish latent RFQT makers

feuGeneA opened the pull request:
## Description

<!--- Describe your changes in detail -->

## Testing instructions

<!--- Please describe how reviewers can test your changes -->

## Types of changes

<!--- What types of changes does your code introduce? Uncomment all the bullets that apply: -->

<!-- * Bug fix (non-breaking change which fixes an issue) -->

<!-- * New feature (non-breaking change which adds functionality) -->

<!-- * Breaking change (fix or feature that would cause existing functionality to change) -->

## Checklist:

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

feuGeneA committed:
asset-swapper: Punish latent RFQT makers

PirosB3 commented on packages/asset-swapper/src/utils/rfq_maker_blacklist.ts:
```diff
@@ -0,0 +1,38 @@
+    private static readonly _msPerMinute = 1000 * 60;
+    private readonly _makerTimeoutStreakLength: { [makerUrl: string]: number } = {};
+    private readonly _makerBlacklistedUntilDate: { [makerUrl: string]: number } = {};
+    constructor(private readonly _blacklistDurationMinutes: number, private readonly _timeoutStreakThreshold: number) {}
```
why do constructor args have underscores? I know you are taking a shortcut here which I took in the past, but I was suggested to simply set up constructor the old fashioned way

feuGeneA replied:
the underscores are there because the linter required me to put them there, because they're not just function arguments, they're also declarations of member variables.

PirosB3 replied:
right, could we use the old fashioned way instead? or do u think it's not worth it? underscore arguments usually are an alias for unused arguments

PirosB3 commented on packages/asset-swapper/src/utils/rfq_maker_blacklist.ts:
```diff
@@ -0,0 +1,38 @@
+ */
+export class RfqMakerBlacklist {
+    // tslint:disable-next-line:custom-no-magic-numbers
+    private static readonly _msPerMinute = 1000 * 60;
```
Can you replace `1000` with `ONE_SECOND`?

PirosB3 replied:
Should this live in this class? or could it be in the same file of `ONE_SECOND` or another const file?

feuGeneA replied:
Addressed in https://github.com/0xProject/0x-monorepo/pull/2706/commits/79e0a9ef3701f8bdb2e28dd4a6b83bce88019472

PirosB3 commented on packages/asset-swapper/src/utils/rfq_maker_blacklist.ts:
```diff
@@ -0,0 +1,38 @@
+    }
+    public isMakerBlacklisted(makerUrl: string): boolean {
+        if (this._makerBlacklistedUntilDate.hasOwnProperty(makerUrl)) {
+            if (this._makerBlacklistedUntilDate[makerUrl] > Date.now()) {
```
what is the difference between `new Date().getTime()` and `Date.now()`? we usually use the former afaik

feuGeneA replied:
No difference in effect, really.  I prefer `Date.now()`, partly because in theory it's less overhead (no constructing a `Date` object), but more importantly because it's easy to tell at a glance what the intention is, without needing to remember that `new Date()` constructs an object set to `"now"`.

PirosB3 commented on packages/asset-swapper/src/utils/rfq_maker_blacklist.ts:
```diff
@@ -0,0 +1,38 @@
+            this._makerTimeoutStreakLength[makerUrl] = 0;
+        }
+        if (didTimeout) {
+            this._makerTimeoutStreakLength[makerUrl] += 1;
```
`this._makerTimeoutStreakLength[makerUrl] = (this._makerTimeoutStreakLength[makerUrl] || 0) + 1` could allow you to omit L12-14

feuGeneA replied:
I loved the suggestion at first.  But then, when I wrote it up, I decided that I prefer to keep the simple `...Streak... += 1;`, since that's such a key part of the mechanism here, and I'd like to keep it as simple and straightforward as possible.

PirosB3 reviewed (changes requested):
Can we write some tests for the new class?

feuGeneA committed:
Add and use constant ONE_MINUTE_MS

Addresses review comments
https://github.com/0xProject/0x-monorepo/pull/2706#discussion_r489610605
and
https://github.com/0xProject/0x-monorepo/pull/2706#discussion_r489614731

feuGeneA committed:
Streamline isMakerBlacklisted() logic

feuGeneA committed:
Blacklist when threshhold reached, not surpassed

feuGeneA committed:
Add tests of class RfqMakerBlacklist

Addresses review comment https://github.com/0xProject/0x-monorepo/pull/2706#pullrequestreview-489852363
